File: ddss/orm.py
import asyncio
from collections import defaultdict
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession
from sqlalchemy.exc import IntegrityError
from sqlalchemy import Integer, Text
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase):
    pass


class Facts(Base):
    __tablename__ = "facts"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    data: Mapped[str] = mapped_column(Text, unique=True, nullable=False)


class Ideas(Base):
    __tablename__ = "ideas"
    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    data: Mapped[str] = mapped_column(Text, unique=True, nullable=False)


async def initialize_database(addr: str) -> tuple[AsyncEngine, async_sessionmaker[AsyncSession]]:
    engine = create_async_engine(addr)
    session = async_sessionmaker(engine)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)
    return engine, session


async def insert_or_ignore(sess: AsyncSession, model, data, locks=defaultdict(lambda: asyncio.Lock())) -> None:
    async with locks[id(sess.bind)]:
        try:
            async with sess.begin_nested():
                sess.add(model(data=data))
                await sess.flush()
        except IntegrityError:
            pass

File: ddss/orm.ts
import {
    DataTypes,
    Model,
    Sequelize,
    type CreationOptional,
    type InferAttributes,
    type InferCreationAttributes,
    type ModelStatic,
} from "sequelize";

class Fact extends Model<InferAttributes<Fact>, InferCreationAttributes<Fact>> {
    declare id: CreationOptional<number>;
    declare data: string;
}

class Idea extends Model<InferAttributes<Idea>, InferCreationAttributes<Idea>> {
    declare id: CreationOptional<number>;
    declare data: string;
}

export { Fact, Idea };

export async function initializeDatabase(addr: string): Promise<Sequelize> {
    let sequelizeAddr = addr;
    if (addr.startsWith("sqlite:///")) {
        sequelizeAddr = `sqlite:${addr.replace("sqlite:///", "")}`;
    }

    const sequelize = new Sequelize(sequelizeAddr, { logging: false });

    Fact.init(
        {
            id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
            data: { type: DataTypes.TEXT, unique: true, allowNull: false },
        },
        { sequelize, tableName: "facts", timestamps: false },
    );

    Idea.init(
        {
            id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
            data: { type: DataTypes.TEXT, unique: true, allowNull: false },
        },
        { sequelize, tableName: "ideas", timestamps: false },
    );

    await sequelize.sync();
    return sequelize;
}

export async function insertOrIgnore(model: typeof Fact | typeof Idea, data: string): Promise<void> {
    await model.bulkCreate([{ data }], {
        ignoreDuplicates: true,
    });
}
